rbac/default-role-manager/role_manager.go

```diff
@@ -46,12 +48,41 @@ func newRole(name string) *Role {
 	r.matchedBy = &sync.Map{}
 	r.linkConditionFuncMap = &sync.Map{}
 	r.linkConditionFuncParamsMap = &sync.Map{}
+	r.level = -1
 	return &r
 }
 
+func (r *Role) updateLevel() {
+	maxLevel := 0
+	hasParent := false
+	r.users.Range(func(_, value interface{}) bool {
+		hasParent = true
+		parentRole := value.(*Role)
+		if parentRole.level >= 0 && parentRole.level+1 > maxLevel {
+			maxLevel = parentRole.level + 1
+		}
+		return true
+	})
+
+	newLevel := 0
+	if hasParent {
+		newLevel = maxLevel
+	}
+
+	if newLevel != r.level {
+		r.level = newLevel
+		r.roles.Range(func(_, value interface{}) bool {
+			role := value.(*Role)
+			role.updateLevel()
+			return true
+		})
+	}
+}
+
```

Comment:
Consider streamlining redundant local variables and overall processes
example:
```go
func (r *Role) updateLevel() {
	maxLevel := 0
	r.users.Range(func(_, value interface{}) bool {
		if parentRole := value.(*Role); parentRole.level >= 0 && parentRole.level+1 > maxLevel {
			maxLevel = parentRole.level + 1
		}
		return true
	})
	if r.level == maxLevel {
		return
	}
	r.level = maxLevel
	r.roles.Range(func(_, value interface{}) bool {
		value.(*Role).updateLevel()
		return true
	})
}
```

Also, what if the role map has a ring?